### PR TITLE
Add `ToString` filter

### LINKED ISSUE #98: ToString filter missing?
### Feature Request

Laminas-filter component has [ToFloat](/laminas/laminas-filter/blob/2.32.x/src/ToFloat.php), [ToInt](/laminas/laminas-filter/blob/2.32.x/src/ToInt.php) and [ToNull](/laminas/laminas-filter/blob/2.32.x/src/ToFloat.php) filters but `ToString` is missing.

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

This filter should be similar as `ToInt`:

```php
class ToString extends AbstractFilter
{
    public function filter($value)
    {
        if (
            is_scalar($value)
            || $value === null
            || $value instanceof Stringable
        ) {
            return (string)  $value;
        }

        return $value;
    }
}
```

Your opinions?

### PR DESCRIPTION
**Pull request by mlussi90**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

#98 Adding a ToString Filter according to the requirements from @vaclavvanik

### CODE REVIEW

File: src/ToString.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\Filter;
+
+use function is_scalar;
+
+class ToString extends AbstractFilter
```
Comment (gsteel) [inline src/ToString.php]:
```suggestion
final class ToString implements FilterInterface
```

File: src/ToString.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    /**
+     * Returns (string) $value
+     *
+     * If the value provided is non-scalar, the value will remain unfiltered
+     *
+     * @param mixed $value
+     * @psalm-return ($value is scalar ? string : mixed)
+     */
+    public function filter($value): mixed
```
Comment (gsteel) [inline src/ToString.php]:
```suggestion
     * @return ($value is scalar ? string : mixed)
     */
    public function filter(mixed $value): mixed
```

File: src/ToString.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        if (! is_scalar($value)) {
+            return $value;
+        }
+
+        return (string) $value;
+    }
```
Comment (gsteel) [inline src/ToString.php:30]:
```suggestion
    }
    
    public function __invoke(mixed $value): mixed
    {
        return $this->filter($value);
    }
```

File: test/ToStringTest.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+namespace LaminasTest\Filter;
+
+use Laminas\Filter\ToString;
+use PHPUnit\Framework\Attributes\DataProvider;
+use PHPUnit\Framework\TestCase;
+use stdClass;
+
+class ToStringTest extends TestCase
+{
+    /** @return list<array{0: mixed}> */
```
Comment (gsteel) [inline test/ToStringTest.php]:
```suggestion
    /** @return list<array{0: mixed, 1: string}> */
```

File: src/ToString.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+     * Returns (string) $value
+     *
+     * If the value provided is non-scalar, the value will remain unfiltered
+     *
+     * @param mixed $value
+     * @psalm-return ($value is scalar ? string : mixed)
+     */
+    public function filter($value): mixed
+    {
+        if (! is_scalar($value)) {
```
Comment (gsteel) [inline src/ToString.php]: This condition should include `Stringable`

File: test/ToStringTest.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    public static function returnBasicDataProvider(): array
```
Comment (gsteel) [inline test/ToStringTest.php:16]: Data providers always appear as unused by Psalm when only referenced by the `DataProvider` attribute.

Once all other psalm issues are fixed, these 'PossiblyUnusedMethod' issues can be added to the baseline with `psalm --set-baseline=psalm-baseline.xml`

File: test/ToStringTest.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
+<?php
```
Comment (gsteel) [inline test/ToStringTest.php:1]: Will also need a test to cover added `__invoke` method and data provider cases covering `Stringable` objects

**Review by gsteel — CHANGES_REQUESTED** [review]
Thanks @mlussi90 

The `ToString` filter will need to be added to the default set of filters in `FilterPluginManager` - additionally, the v2 docs will need updating 👍

**Comment by mlussi90** [comment]
i tried to implement all your requested changes, can you take a look?

File: src/FilterPluginManager.php [hunk 8]
```diff
@@ -32,10 +32,12 @@ class FilterPluginManager extends AbstractPluginManager
...
-        'int'  => ToInt::class,
-        'Int'  => ToInt::class,
-        'null' => ToNull::class,
-        'Null' => ToNull::class,
+        'int'    => ToInt::class,
+        'Int'    => ToInt::class,
+        'null'   => ToNull::class,
+        'Null'   => ToNull::class,
+        'string' => ToString::class,
+        'String' => ToString::class,
```
Comment (gsteel) [inline src/FilterPluginManager.php]: I'd rather not see any new aliases please - It's better that users get into the habit of referring to filters by the FQCN, i.e. `\Laminas\Filter\ToString::class`

File: src/FilterPluginManager.php [hunk 9]
```diff
@@ -158,6 +160,9 @@ class FilterPluginManager extends AbstractPluginManager
         'tonull'                     => ToNull::class,
         'toNull'                     => ToNull::class,
         'ToNull'                     => ToNull::class,
+        'tostring'                   => ToString::class,
+        'toString'                   => ToString::class,
+        'ToString'                   => ToString::class,
```
Comment (gsteel) [inline src/FilterPluginManager.php]: Same comment about aliases here too. Please remove.

File: src/FilterPluginManager.php [hunk 10]
```diff
@@ -362,6 +367,7 @@ class FilterPluginManager extends AbstractPluginManager
         Inflector::class                   => InvokableFactory::class,
         ToInt::class                       => InvokableFactory::class,
         ToFloat::class                     => InvokableFactory::class,
+        ToString::class                    => InvokableFactory::class,
```
Comment (gsteel) [inline src/FilterPluginManager.php:365]: This should be all we need 👍

File: test/ToStringTest.php [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+        self::assertSame($output, $filter->filter($input));
+    }
+
+    #[DataProvider('returnBasicDataProvider')]
+    public function testInvoke(mixed $input, string $output): void
+    {
+        $filter = new ToString();
+
+        self::assertSame($output, $filter($input));
```
Comment (gsteel) [inline test/ToStringTest.php]:
```suggestion
        self::assertSame($output, $filter->__invoke($input));
```

Sorry - a bit nit picky, but explicitly calling `__invoke` will ensure that the call cannot be interpreted any other way.

**Review by gsteel — CHANGES_REQUESTED** [review]
Thanks @mlussi90 - Nearly there…

You can fix coding style problems by running `vendor/bin/phpcbf` and psalm issues with the tests can be silenced with `vendor/bin/psalm --set-baseline=psalm-baseline.xml`

**Review by gsteel — CHANGES_REQUESTED** [review]
Thank you @mlussi90 

In order to get CI green for Psalm, you'll need to refresh the baseline - this sort of thing can happen, particularly on Mac, so it's not your fault!

Try running `vendor/bin/psalm --update-baseline --no-cache --threads=1` and verify there are no more problems with `vendor/bin/psalm --no-cache --threads=1`, then commit the altered baseline.

The BC check failure can be ignored - it's complaining about the altered default value for plugin manager factories.

**Comment by mlussi90** [comment]
@gsteel thank you very much for your help, the checks are looking better now i think

**Review by gsteel — APPROVED** [review]
Thanks @mlussi90 

I'll merge and release shortly, once I've got the docs organised 👍

**Pull request merged**